packages/playground/blueprints/src/lib/compile.ts

```diff
@@ -171,33 +191,81 @@ export function compileBlueprint(
 			networking: blueprint.features?.networking ?? false,
 		},
 		run: async (playground: UniversalPHP) => {
+			const progressStatus = new Map<StepDefinition, Promise<void>>();
+			const progressPromises = new Map<
+				StepDefinition,
+				{
+					resolve: (result: any) => void;
+					reject: (error: any) => void;
+				}
+			>();
+			for (const { step } of compiled) {
+				const promise = new Promise<void>((resolve, reject) => {
+					progressPromises.set(step, { resolve, reject });
+				});
+				progressStatus.set(step, promise);
+				onStepProgress(promise, step);
+			}
+
 			try {
 				// Start resolving resources early
-				for (const { resources } of compiled) {
+				for (const { resources, step } of compiled) {
+					const stepPromise = progressPromises.get(step);
 					for (const resource of resources) {
 						resource.setPlayground(playground);
 						if (resource.isAsync) {
-							resource.resolve();
+							resource.resolve().catch(function (error) {
+								stepPromise?.reject(error);
+							});
 						}
 					}
 				}
 
+				/**
+				 * When a compiled step's `run` method is called, it
+				 * uses `resolveArguments` to await dependent resources.
+				 */
 				for (const { run, step } of compiled) {
-					const result = await run(playground);
-					onStepCompleted(result, step);
+					const stepPromise = progressPromises.get(step);
+					try {
+						const result = await run(playground);
+						stepPromise?.resolve(result);
+						onStepCompleted(result, step);
+					} catch (error) {
+						stepPromise?.reject(error);
```

Comment:
I wonder whether the failure consequences should be possible to specify per step type, e.g. `rm` could default to triggering a warning but not stopping the execution, while `defineWpConfigConsts` would default to throwing an error and stopping the execution.

There seem to be a hierarchy of control here:

1. Consumer has a preference. If missing, fall back to...
2. Step has a preference. If missing, fall back to...
3. Blueprint engine has a preference – that's implicit from the code structure here

It would be nice to stop the execution as soon as we know the failure is fatal. Imagine the first step of a work-intense Blueprint fails and we already know this is irrecoverable – stopping right then and there would spare the user 30s of waiting before learning about the error.

Code-wise, it could mean a logic similar to this:

```suggestion
						stepPromise?.reject(error);
						if( isFailureFatal( step, error ) ) {
							throw e;
						}
```

Now, the `isFailureFatal` would need to know about both the step preference and the user preference, which means we need to provide developers with a way of overriding the default behavior on a granular, per-step basis.

This PR introduces a decision point in the `onStepsCompleted` callback registered by the application calling the `compileBlueprint` function. I think it would be more useful to make that decision inside of a Blueprint. Imagine a live product demo. The `installPlugin` installing the demoed plugin is indispensable, but the next one installing a dev utility plugin can likely be skipped if it fails.

Technically, this could mean introducing a new step-level property:

```diff
export type StepDefinition = Step & {
	progress?: {
		weight?: number;
		caption?: string;
	};
+	onFailure: 'abort' | 'skip';
};
```

This would enable writing blueprints such as...

```json
{
	"steps": [
		{
			"step": "installPlugin",
			"pluginZipFile": {
				"resource": "wordpress.org/plugins",
				"slug": "my-product"
			},
			"onFailure": "abort"
		},
		{
			"step": "installPlugin",
			"pluginZipFile": {
				"resource": "wordpress.org/plugins",
				"slug": "devtools"
			},
			"onFailure": "skip"
		}
}
```

The default value could be `abort`. Each step handler would be able to override it, for example like this:

```ts
/**
 * Removes a file at the specified path.
 */
export const rm: StepHandler<RmStep> = async (playground, { path }) => {
	await playground.unlink(path);
};

import { defaultStepFailureHandlers } from '.';
defaultStepFailureHandlers.set( 'rm', 'continue' );
```

The `compileStep()` function could then output the desired failure behavior:

```ts

function compileStep<S extends StepDefinition>(
	step: S,
	{
		semaphore,
		rootProgressTracker,
		totalProgressWeight,
	}: CompileStepArgsOptions
): { run: CompiledStep; step: S; resources: Array<Resource> } {
	return {
		run,
		step,
		resources,
		onFailure:  'onFailure' in step ? step.onFailure : defaultStepFailureHandlers.get( step.step ) ?? 'abort'
	};
```

And this loop could handle it as follows:

```ts
				for (const { run, step, onFailure } of compiled) {
					const stepPromise = progressPromises.get(step);
					try {
						onStepCompleted(result, step);
					} catch(e) {
						if( onFailure === 'abort' ) {
							throw e;
						}
					}
```

Reply:
I wonder how the ideas above resonate with you @eliot-akira @dmsnell @bgrgicak

Reply:
I think it fits within the idea I had, which was to expose each to the application and let it decide.

we could provide default behaviors, which I think should be rare, and then if someone in userspace adds their own handler for the step, then they could take their own action.

or, we could expose our default handling to the callback and only if they implement the callback would we use theirs instead.

Reply:
I like this approach, it would at least help with development and while creating blueprints.

Personally, I love opinionated code that I can adjust, so I would go with having defaults for each step, but let users override them. Maybe even include a global override for all steps.

Reply:
Thanks @adamziel for the detailed review. I'm starting to get a feel for the public interface that's shaping up.

I like the idea of adding to the `StepDefinition` a new property `onFailure`, with value `abort` or `skip`. That addresses the original issue that motivated this PR, which is to allow some plugins to be skipped if install fails, while still keeping plugins required by default (error on install fail).

It makes sense for each step handler to have a default failure mode (`abort` or `skip`), which can be overridden by the user.

I'll get started on implementing the suggestions, and will either resolve each reviewed point or reply with any questions.

Reply:
Thank you so much @eliot-akira, you are the best!